db/db_logical_block_size_cache_test.cc

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) 2020-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under both the GPLv2 (found in the
+// COPYING file in the root directory) and Apache 2.0 License
+// (found in the LICENSE.Apache file in the root directory).
+
+#include "test_util/testharness.h"
+
+#ifdef OS_LINUX
+#include "env/io_posix.h"
+#include "rocksdb/db.h"
+#include "rocksdb/env.h"
+
+namespace ROCKSDB_NAMESPACE {
+class EnvWithCustomLogicalBlockSizeCache : public EnvWrapper {
+ public:
+  EnvWithCustomLogicalBlockSizeCache(Env* env, LogicalBlockSizeCache* cache)
+      : EnvWrapper(env), cache_(cache) {}
+
+  Status RegisterDbPaths(const std::vector<std::string>& paths) override {
+    return cache_->RefAndCacheLogicalBlockSize(paths);
+  }
+
+  Status UnregisterDbPaths(const std::vector<std::string>& paths) override {
+    cache_->UnrefAndTryRemoveCachedLogicalBlockSize(paths);
+    return Status::OK();
+  }
+
+ private:
+  LogicalBlockSizeCache* cache_;
+};
+
+class DBLogicalBlockSizeCacheTest : public testing::Test {
+ public:
+  DBLogicalBlockSizeCacheTest()
+      : dbname_(test::PerThreadDBPath("logical_block_size_cache_test")),
+        data_path_0_(dbname_ + "/data_path_0"),
+        data_path_1_(dbname_ + "/data_path_1"),
+        cf_path_0_(dbname_ + "/cf_path_0"),
+        cf_path_1_(dbname_ + "/cf_path_1"),
+        kDirFds({
+          {dbname_, 0},
+          {data_path_0_, 10},
+          {data_path_1_, 11},
+          {cf_path_0_, 20},
+          {cf_path_1_, 21}
+        }) {
+    auto get_fd_block_size = [&](int fd) {
+      ncall_++;
+      return fd;
+    };
+    auto get_dir_block_size = [&](const std::string& dir, size_t* size) {
+      ncall_++;
+      *size = kDirFds.at(dir);
+      return Status::OK();
+    };
+    cache_.reset(new LogicalBlockSizeCache(
+        get_fd_block_size, get_dir_block_size));
+    env_.reset(new EnvWithCustomLogicalBlockSizeCache(
+        Env::Default(), cache_.get()));
+  }
+
+ protected:
+  std::string dbname_;
+  std::string data_path_0_;
+  std::string data_path_1_;
+  std::string cf_path_0_;
+  std::string cf_path_1_;
+
+  const std::map<std::string, int> kDirFds;
+
+  int ncall_ = 0;
+  std::unique_ptr<LogicalBlockSizeCache> cache_;
+  std::unique_ptr<Env> env_;
+};
+
+TEST_F(DBLogicalBlockSizeCacheTest, OpenClose) {
+  // Tests that Open will cache the logical block size for data paths,
+  // and Close will remove the cached sizes.
+  Options options;
+  options.create_if_missing = true;
+  options.env = env_.get();
+  options.db_paths = {{data_path_0_, 2048}, {data_path_1_, 2048}};
+
+  for (int i = 0; i < 2; i++) {
+    ncall_ = 0;
+    DB* db;
+    if (!i) {
+      printf("Open\n");
+      ASSERT_OK(DB::Open(options, dbname_, &db));
+    } else {
+      printf("OpenForReadOnly\n");
+      ASSERT_OK(DB::OpenForReadOnly(options, dbname_, &db));
+    }
+    // Logical block size of data paths are cached during Open.
+    ASSERT_EQ(2, ncall_);
```

Comment:
After second thought, for this test, it's sufficient to check what directories are cached. Checking `ncall_` to verify which directory is cached is indirect, instead, I added `Size` and `Contains` to the cache to check the content of the cache directly. 

From DB's perspective, if I can verify that a directory is cached, then I don't need to verify whether `GetLogicalBlockSize` on that cached directory will increase `ncall_` because that's ensured by the cache's implementation, and is already verified in the cache's unit test in `io_posix_test.cc`.